File: tape/sdk/java/src/test/java/dev/tape/embedded/chaos/ChaosTest.java
package dev.tape.embedded.chaos;

import dev.tape.embedded.CompensationResult;
import dev.tape.embedded.Connector;
import dev.tape.embedded.DispatchResult;
import dev.tape.embedded.ObservationResult;
import dev.tape.embedded.Reactors;
import dev.tape.embedded.Schema;
import dev.tape.embedded.SqliteDataSource;
import dev.tape.embedded.TapeSessionService;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static dev.tape.embedded.Schema.EffectRecord;
import static dev.tape.embedded.Schema.ObligationRecord;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Embedded-tier chaos tests — mirrors
 * {@code tape/sdk/python-adk/tests/test_chaos.py}. Proves the same
 * invariants the Python suite proves, plus the strict-faults
 * false-positive guard.
 */
class ChaosTest {

    private SqliteDataSource ds;
    private TapeSessionService svc;

    @BeforeEach void setup() {
        ds = new SqliteDataSource();
        svc = new TapeSessionService(ds);
    }

    @AfterEach void teardown() { ds.shutdown(); }

    /** A trivial idempotent ledger connector — fixture analogue of the
     *  Python {@code _LedgerConnector}. CONFIRMED on dispatch; mirrors the
     *  business_key dedupe a real bank's API provides. */
    static final class LedgerConnector implements Connector {
        final String name;
        final Map<String, String> ledger = new LinkedHashMap<>();
        long delayMs = 0L;

        LedgerConnector() { this("bank.wire"); }
        LedgerConnector(String name) { this.name = name; }

        @Override public String name() { return name; }

        @Override public DispatchResult dispatch(EffectRecord effect) throws Exception {
            if (delayMs > 0) Thread.sleep(delayMs);
            String bk = effect.businessKey() != null && !effect.businessKey().isEmpty()
                ? effect.businessKey() : effect.idempotencyKey();
            String wid = ledger.computeIfAbsent(bk,
                k -> String.format("w-%04d", ledger.size()));
            return DispatchResult.confirmed(wid,
                "{\"wire_id\":\"" + wid + "\"}");
        }

        @Override public ObservationResult observe(EffectRecord effect) {
            String bk = effect.businessKey() != null && !effect.businessKey().isEmpty()
                ? effect.businessKey() : effect.idempotencyKey();
            if (ledger.containsKey(bk)) {
                return ObservationResult.confirmed(ledger.get(bk), null);
            }
            return ObservationResult.absent();
        }

        @Override public CompensationResult compensate(ObligationRecord obligation) {
            return CompensationResult.compensated(null);
        }
    }

    private static EffectRecord syntheticEffect(String tool, String connector, String key) {
        return new EffectRecord(
            "a", "u", "s", "k-" + tool, "inv-1",
            0, tool, 0,
            EffectRecord.PENDING, EffectRecord.NON_IDEMPOTENT, EffectRecord.OUTBOX,
            "bk-" + tool, connector, null,
            0, 0L, null, 0L, null,
            "{}", null, null, 0L);
    }

    // ── ChaosConnector: the fault mechanism, in isolation ────────────────

    @Test
    void loseAckFlipsConfirmedToUnknown() throws Exception {
        LedgerConnector inner = new LedgerConnector();
        ChaosConnector wrapped = new ChaosConnector(
            inner,
            List.of(Chaos.loseAck("bank.wire", "", 1.0)),
            new java.util.Random(0L));

        EffectRecord effect = new EffectRecord(
            "a", "u", "s", "k", "inv-1",
            0, "wire", 0,
            EffectRecord.PENDING, EffectRecord.NON_IDEMPOTENT, EffectRecord.OUTBOX,
            "bk-1", "bank.wire", null,
            0, 0L, null, 0L, null,
            "{}", null, null, 0L);
        DispatchResult result = wrapped.dispatch(effect);
        assertEquals(DispatchResult.UNKNOWN, result.status());
        // The inner call did land (the wrapper's contract).
        assertEquals(1, inner.ledger.size());
        assertTrue(inner.ledger.containsKey("bk-1"));
    }

    @Test
    void delayConnectorBlocksDispatch() throws Exception {
        LedgerConnector inner = new LedgerConnector();
        ChaosConnector wrapped = new ChaosConnector(
            inner,
            List.of(Chaos.delayConnector("bank.wire", 80L)),
            new java.util.Random(0L));

        EffectRecord effect = new EffectRecord(
            "a", "u", "s", "k", "inv-1",
            0, "wire", 0,
            EffectRecord.PENDING, EffectRecord.IDEMPOTENT, EffectRecord.INLINE,
            null, "bank.wire", null,
            0, 0L, null, 0L, null,
            "{}", null, null, 0L);
        long t0 = System.nanoTime();
        wrapped.dispatch(effect);
        long elapsedMs = (System.nanoTime() - t0) / 1_000_000;
        assertTrue(elapsedMs >= 70, "honoured the delay; was " + elapsedMs + "ms");
    }

    @Test
    void toolScopedFaultOnlyFiresOnMatchingTool() throws Exception {
        LedgerConnector inner = new LedgerConnector();
        ChaosConnector wrapped = new ChaosConnector(
            inner,
            List.of(Chaos.loseAck("", "wire", 1.0)),
            new java.util.Random(0L));

        DispatchResult rWire = wrapped.dispatch(
            syntheticEffect("wire", "bank.wire", "bk-wire"));
        DispatchResult rPost = wrapped.dispatch(
            syntheticEffect("post_gl", "bank.wire", "bk-post"));
        assertEquals(DispatchResult.UNKNOWN, rWire.status(),
            "tool matches → fault fires");
        assertEquals(DispatchResult.CONFIRMED, rPost.status(),
            "tool doesn't match → passthrough");
    }

    // ── strict_faults: the silent-skip false-positive guard ─────────────

    @Test
    void strictFaultsFailsOnMissingConnector() throws Exception {
        Scenario scen = Chaos.scenario(
            "missing-target",
            List.of(Chaos.loseAck("bank.wire", "")),
            List.of(Invariants.NO_STUCK_OBLIGATIONS));
        ChaosReport report = Chaos.run(scen, svc, ds,
            new HashMap<>(),  // empty — bank.wire missing
            c -> {});
        assertFalse(report.passed());
        assertTrue(report.invariantResults().stream()
            .anyMatch(r -> r.toString().contains("strict_faults")));
    }

    @Test
    void strictFaultsOffAllowsSkip() throws Exception {
        Scenario scen = new Scenario(
            "optional-target",
            List.of(Chaos.loseAck("bank.wire", "")),
            List.of(Invariants.NO_STUCK_OBLIGATIONS),
            0L,
            false);   // strict_faults off
        ChaosReport report = Chaos.run(scen, svc, ds,
            new HashMap<>(), c -> {});
        assertTrue(report.passed());
        assertTrue(report.notes().stream()
            .anyMatch(n -> n.contains("not in `connectors` dict")));
    }

    // ── invariants: read the embedded tables ─────────────────────────────

    @Test
    void noStuckObligationsPassesOnCleanStore() throws Exception {
        Scenario scen = Chaos.scenario("smoke",
            List.of(),
            List.of(Invariants.NO_STUCK_OBLIGATIONS));
        ChaosReport report = Chaos.run(scen, svc, ds,
            new HashMap<>(), c -> {});
        assertTrue(report.passed(), report.toString());
    }

    @Test
    void noStuckObligationsFailsWhenOneIsStuck() throws Exception {
        ObligationRecord ob = svc.registerCompensation(
            "a", "u", "s", "", "ek", "reverse_wire", null, null, 1);
        svc.resolveObligation(ob.seq(), ObligationRecord.STUCK, null);

        Scenario scen = Chaos.scenario("stuck",
            List.of(),
            List.of(Invariants.NO_STUCK_OBLIGATIONS));
        ChaosReport report = Chaos.run(scen, svc, ds,
            new HashMap<>(), c -> {});
        assertFalse(report.passed());
        assertTrue(report.invariantResults().stream()
            .anyMatch(r -> r.toString().toLowerCase().contains("stuck")));
    }

    @Test
    void exactlyOneInvariant() throws Exception {
        EffectRecord e = svc.beginEffect(
            "a", "u", "s", "inv-1", 0, "wire", 0,
            null, null,
            EffectRecord.NON_IDEMPOTENT, EffectRecord.OUTBOX,
            "bk-1", "bank.wire");
        svc.completeEffect("a", "u", "s", e.idempotencyKey(),
            EffectRecord.CONFIRMED, "{\"id\":\"1\"}", null);

        Scenario scen = Chaos.scenario("one",
            List.of(),
            List.of(Invariants.exactlyOneByConnector("bank.wire")));
        ChaosReport report = Chaos.run(scen, svc, ds,
            new HashMap<>(), c -> {});
        assertTrue(report.passed(), report.toString());
    }

    // ── end-to-end: lose_ack → reconcile loop drives UNKNOWN to CONFIRMED ─

    @Test
    void loseAckE2EWithReconciler() throws Exception {
        EffectRecord eff = svc.beginEffect(
            "a", "u", "s", "inv-1", 0, "wire", 0,
            null, null,
            EffectRecord.NON_IDEMPOTENT, EffectRecord.OUTBOX,
            "bk-1", "bank.wire");
        final LedgerConnector bank = new LedgerConnector();

        Scenario scen = Chaos.scenario(
            "unknown-then-reconcile",
            List.of(Chaos.loseAck("bank.wire", "", 1.0)),
            List.of(
                Invariants.NO_STUCK_OBLIGATIONS,
                Invariants.exactlyOneByConnector("bank.wire")));

        Map<String, Connector> connectors = new HashMap<>();
        connectors.put("bank.wire", bank);

        ChaosReport report = Chaos.run(scen, svc, ds, connectors,
            wrapped -> {
                // Tick 1: dispatch (gets UNKNOWN due to lose_ack).
                List<Reactors.ReactorAction> r1 = Reactors.dispatchOutboxOnce(
                    svc, wrapped, "d-1");
                assertTrue(r1.stream().anyMatch(a -> "unknown".equals(a.outcome())),
                    "expected UNKNOWN outcome: " + r1);
                // The bank's ledger has exactly one wire (inner call landed).
                assertEquals(1, bank.ledger.size());
                // Tick 2: reconciler observes the unwrapped bank and resolves.
                Map<String, Connector> bare = new HashMap<>();
                bare.put("bank.wire", bank);
                List<Reactors.ReactorAction> r2 = Reactors.reconcileOnce(
                    svc, bare);
                assertTrue(r2.stream().anyMatch(a -> "confirmed".equals(a.outcome())),
                    "expected CONFIRMED resolution: " + r2);
            });
        assertTrue(report.passed(), report.toString());
        // And exactly one wire on the bank's side: the property the whole
        // contract exists to enforce.
        assertEquals(1, bank.ledger.size());
    }

    // ── invariant API uniformity ─────────────────────────────────────────

    @Test
    void invariantsRejectBothScopesSet() {
        // exactly_one rejects both connector and tool simultaneously.
        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
            () -> Invariants.exactlyOne("bank.wire", "wire"));
        assertTrue(ex.getMessage().contains("connector= or tool="));
    }

    @Test
    void loseAckRejectsBothScopesSet() {
        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
            () -> Chaos.loseAck("bank.wire", "wire", 1.0));
        assertTrue(ex.getMessage().contains("connector= or tool="));
    }
}

File: tape/sdk/go/embedded/chaos_test.go
package embedded

// chaos_test.go — Go port of `tape_adk/tests/test_chaos.py`. Same ten
// tests, same invariant proofs, same strict-faults guard. All run
// against an in-memory SQLite via `modernc.org/sqlite`.

import (
	"context"
	"math/rand/v2"
	"strings"
	"sync"
	"testing"
	"time"
)

// ── a configurable fake ledger connector (test fixture) ──────────────────

// ledgerConnector — a trivial idempotent ledger. CONFIRMED on dispatch;
// mirrors the business_key dedupe a real bank's API provides.
type ledgerConnector struct {
	nameStr string

	mu      sync.Mutex
	ledger  map[string]string // business_key → wire_id
	delayMs int
}

func newLedgerConnector() *ledgerConnector {
	return &ledgerConnector{nameStr: "bank.wire", ledger: map[string]string{}}
}

func (l *ledgerConnector) Name() string { return l.nameStr }

func (l *ledgerConnector) Dispatch(ctx context.Context, e EffectRecord) (DispatchResult, error) {
	if l.delayMs > 0 {
		select {
		case <-time.After(time.Duration(l.delayMs) * time.Millisecond):
		case <-ctx.Done():
			return DispatchResult{}, ctx.Err()
		}
	}
	bk := e.BusinessKey
	if bk == "" {
		bk = e.IdempotencyKey
	}
	l.mu.Lock()
	wid, ok := l.ledger[bk]
	if !ok {
		wid = padWireID(len(l.ledger))
		l.ledger[bk] = wid
	}
	l.mu.Unlock()
	return DispatchResult{Status: "confirmed", ExternalRef: wid,
		Response: map[string]any{"wire_id": wid}}, nil
}

func (l *ledgerConnector) Observe(_ context.Context, e EffectRecord) (ObservationResult, error) {
	bk := e.BusinessKey
	if bk == "" {
		bk = e.IdempotencyKey
	}
	l.mu.Lock()
	wid, ok := l.ledger[bk]
	l.mu.Unlock()
	if ok {
		return ObservationResult{Status: "confirmed", ExternalRef: wid,
			Response: map[string]any{"wire_id": wid}}, nil
	}
	return ObservationResult{Status: "absent"}, nil
}

func (l *ledgerConnector) Compensate(_ context.Context, _ ObligationRecord) (CompensationResult, error) {
	return CompensationResult{Status: "compensated"}, nil
}

func (l *ledgerConnector) ledgerLen() int {
	l.mu.Lock()
	defer l.mu.Unlock()
	return len(l.ledger)
}

// padWireID — "w-0000", "w-0001", … (same shape as Python).
func padWireID(n int) string {
	digits := []byte{'0', '0', '0', '0'}
	for i := 3; i >= 0 && n > 0; i-- {
		digits[i] = byte('0' + n%10)
		n /= 10
	}
	return "w-" + string(digits)
}

// fixedEffect — synthesize an EffectRecord for unit tests that need to
// drive `ChaosConnector` directly without going through the service.
func fixedEffect(tool, bk string) EffectRecord {
	return EffectRecord{
		AppName: "a", UserID: "u", SessionID: "s",
		IdempotencyKey: "k-" + tool, InvocationID: "inv-1",
		DecisionIndex: 0, ToolName: tool, CallIndex: 0,
		Status: "pending", Semantics: "non_idempotent",
		DispatchMode: "outbox",
		BusinessKey:  bk, Connector: "bank.wire",
	}
}

// ── ChaosConnector: the fault mechanism, in isolation ────────────────────

func TestLoseAckFlipsConfirmedToUnknown(t *testing.T) {
	inner := newLedgerConnector()
	f := MustLoseAck("bank.wire", "", 1.0)
	wrapped := NewChaosConnector(inner, []Fault{f}, rand.New(rand.NewPCG(1, 2)))

	effect := fixedEffect("wire", "bk-1")
	result, err := wrapped.Dispatch(context.Background(), effect)
	if err != nil {
		t.Fatalf("dispatch: %v", err)
	}
	if result.Status != "unknown" {
		t.Fatalf("status=%q, want unknown", result.Status)
	}
	// The inner call did land (the wrapper's contract).
	if inner.ledgerLen() != 1 {
		t.Fatalf("ledger=%d, want 1 (inner call must land)", inner.ledgerLen())
	}
}

func TestDelayConnectorBlocksDispatch(t *testing.T) {
	inner := newLedgerConnector()
	wrapped := NewChaosConnector(inner,
		[]Fault{DelayConnector("bank.wire", 80, 0)},
		rand.New(rand.NewPCG(1, 2)))

	effect := EffectRecord{
		AppName: "a", UserID: "u", SessionID: "s",
		IdempotencyKey: "k", InvocationID: "inv-1",
		DecisionIndex: 0, ToolName: "wire", CallIndex: 0,
		Status: "pending", Semantics: "idempotent",
		DispatchMode: "inline", Connector: "bank.wire",
	}
	t0 := time.Now()
	if _, err := wrapped.Dispatch(context.Background(), effect); err != nil {
		t.Fatalf("dispatch: %v", err)
	}
	if d := time.Since(t0); d < 70*time.Millisecond {
		t.Fatalf("elapsed=%v, want >= 70ms (delay honoured)", d)
	}
}

func TestToolScopedFaultOnlyFiresOnMatchingTool(t *testing.T) {
	inner := newLedgerConnector()
	f := MustLoseAck("", "wire", 1.0)
	wrapped := NewChaosConnector(inner, []Fault{f}, rand.New(rand.NewPCG(1, 2)))

	rWire, err := wrapped.Dispatch(context.Background(), fixedEffect("wire", "bk-wire"))
	if err != nil {
		t.Fatalf("dispatch wire: %v", err)
	}
	rPost, err := wrapped.Dispatch(context.Background(), fixedEffect("post_gl", "bk-post"))
	if err != nil {
		t.Fatalf("dispatch post_gl: %v", err)
	}
	if rWire.Status != "unknown" {
		t.Fatalf("wire status=%q, want unknown (tool matches → fault fires)", rWire.Status)
	}
	if rPost.Status != "confirmed" {
		t.Fatalf("post_gl status=%q, want confirmed (tool doesn't match → passthrough)", rPost.Status)
	}
}

// ── strict_faults: the silent-skip false-positive guard ──────────────────

func TestStrictFaultsFailsOnMissingConnector(t *testing.T) {
	scen := Scenario{
		Name:       "missing-target",
		Faults:     []Fault{MustLoseAck("bank.wire", "", 1.0)},
		Invariants: []Invariant{NoStuckObligations},
	}
	svc := newTestSvc(t)
	report, err := RunScenario(context.Background(), scen,
		func(_ context.Context, _ map[string]Connector) error { return nil },
		OpenChaosSessionOpts{Connectors: map[string]Connector{}, Svc: svc})
	if err != nil {
		t.Fatalf("run: %v", err)
	}
	if report.Passed {
		t.Fatalf("expected FAIL, got pass: %v", report)
	}
	foundStrict := false
	for _, r := range report.InvariantResults {
		if r.Name == "strict_faults" {
			foundStrict = true
		}
	}
	if !foundStrict {
		t.Fatalf("expected strict_faults invariant in report, got %+v", report.InvariantResults)
	}
}

func TestStrictFaultsOffAllowsSkip(t *testing.T) {
	scen := Scenario{
		Name:         "optional-target",
		Faults:       []Fault{MustLoseAck("bank.wire", "", 1.0)},
		Invariants:   []Invariant{NoStuckObligations},
		StrictFaults: StrictFaultsOff(),
	}
	svc := newTestSvc(t)
	report, err := RunScenario(context.Background(), scen, nil,
		OpenChaosSessionOpts{Connectors: map[string]Connector{}, Svc: svc})
	if err != nil {
		t.Fatalf("run: %v", err)
	}
	if !report.Passed {
		t.Fatalf("expected pass with StrictFaults=off, got: %v", report)
	}
	foundNote := false
	for _, n := range report.Notes {
		if strings.Contains(n, "not in `connectors` dict") {
			foundNote = true
		}
	}
	if !foundNote {
		t.Fatalf("expected 'not in connectors dict' note, got %+v", report.Notes)
	}
}

// ── invariants: read the embedded tables ─────────────────────────────────

func TestNoStuckObligationsPassesOnCleanStore(t *testing.T) {
	scen := Scenario{
		Name:       "smoke",
		Invariants: []Invariant{NoStuckObligations},
	}
	svc := newTestSvc(t)
	report, err := RunScenario(context.Background(), scen, nil,
		OpenChaosSessionOpts{Connectors: map[string]Connector{}, Svc: svc})
	if err != nil {
		t.Fatalf("run: %v", err)
	}
	if !report.Passed {
		t.Fatalf("expected pass on clean store, got: %v", report)
	}
}

func TestNoStuckObligationsFailsWhenOneIsStuck(t *testing.T) {
	svc := newTestSvc(t)
	ctx := context.Background()
	ob, err := svc.RegisterCompensation(ctx, RegisterCompensationOpts{
		AppName: "a", UserID: "u", SessionID: "s",
		EffectKey: "ek", Kind: "reverse_wire", MaxAttempts: 1,
	})
	if err != nil {
		t.Fatalf("register: %v", err)
	}
	if _, err := svc.ResolveObligation(ctx, ob.Seq, ObligationStatusStuck, nil); err != nil {
		t.Fatalf("resolve: %v", err)
	}

	scen := Scenario{Name: "stuck", Invariants: []Invariant{NoStuckObligations}}
	report, err := RunScenario(ctx, scen, nil,
		OpenChaosSessionOpts{Connectors: map[string]Connector{}, Svc: svc})
	if err != nil {
		t.Fatalf("run: %v", err)
	}
	if report.Passed {
		t.Fatalf("expected FAIL with one stuck obligation, got pass: %v", report)
	}
	foundStuck := false
	for _, r := range report.InvariantResults {
		if strings.Contains(strings.ToLower(r.String()), "stuck") {
			foundStuck = true
		}
	}
	if !foundStuck {
		t.Fatalf("expected an invariant result mentioning 'stuck', got %+v", report.InvariantResults)
	}
}

func TestExactlyOneInvariant(t *testing.T) {
	svc := newTestSvc(t)
	ctx := context.Background()
	e, err := svc.BeginEffect(ctx, BeginEffectOpts{
		AppName: "a", UserID: "u", SessionID: "s", InvocationID: "inv-1",
		DecisionIndex: 0, ToolName: "wire", CallIndex: 0,
		Semantics:    EffectSemanticsNonIdempotent,
		DispatchMode: EffectDispatchOutbox,
		BusinessKey:  "bk-1", Connector: "bank.wire",
	})
	if err != nil {
		t.Fatalf("begin: %v", err)
	}
	if _, err := svc.CompleteEffect(ctx, "a", "u", "s", e.IdempotencyKey,
		EffectStatusConfirmed, map[string]any{"id": "1"}, nil); err != nil {
		t.Fatalf("complete: %v", err)
	}

	scen := Scenario{
		Name:       "one",
		Invariants: []Invariant{MustExactlyOne("bank.wire", "")},
	}
	report, err := RunScenario(ctx, scen, nil,
		OpenChaosSessionOpts{Connectors: map[string]Connector{}, Svc: svc})
	if err != nil {
		t.Fatalf("run: %v", err)
	}
	if !report.Passed {
		t.Fatalf("expected pass with exactly one CONFIRMED, got: %v", report)
	}
}

// ── end-to-end: lose_ack → reconcile loop drives an UNKNOWN to CONFIRMED ─

func TestLoseAckE2EWithReconciler(t *testing.T) {
	svc := newTestSvc(t)
	ctx := context.Background()
	if _, err := svc.BeginEffect(ctx, BeginEffectOpts{
		AppName: "a", UserID: "u", SessionID: "s", InvocationID: "inv-1",
		DecisionIndex: 0, ToolName: "wire", CallIndex: 0,
		Semantics:    EffectSemanticsNonIdempotent,
		DispatchMode: EffectDispatchOutbox,
		BusinessKey:  "bk-1", Connector: "bank.wire",
	}); err != nil {
		t.Fatalf("begin: %v", err)
	}
	bank := newLedgerConnector()

	scen := Scenario{
		Name:   "unknown-then-reconcile",
		Faults: []Fault{MustLoseAck("bank.wire", "", 1.0)},
		Invariants: []Invariant{
			NoStuckObligations,
			MustExactlyOne("bank.wire", ""),
		},
	}

	body := func(ctx context.Context, wrapped map[string]Connector) error {
		// Tick 1: dispatch gets UNKNOWN.
		r1, err := DispatchOutboxOnce(ctx, svc, wrapped, DispatchOutboxOpts{Claimer: "d-1"})
		if err != nil {
			return err
		}
		if !anyOutcome(r1, "unknown") {
			t.Fatalf("expected 'unknown' outcome, got %+v", r1)
		}
		// The bank's ledger has exactly one wire (the inner call landed).
		if bank.ledgerLen() != 1 {
			t.Fatalf("bank=%d, want 1", bank.ledgerLen())
		}
		// Tick 2: reconcile (unwrapped bank → CONFIRMED).
		r2, err := ReconcileOnce(ctx, svc, map[string]Connector{"bank.wire": bank}, ReconcileOpts{})
		if err != nil {
			return err
		}
		if !anyOutcome(r2, "confirmed") {
			t.Fatalf("expected 'confirmed' outcome, got %+v", r2)
		}
		return nil
	}

	report, err := RunScenario(ctx, scen, body, OpenChaosSessionOpts{
		Connectors: map[string]Connector{"bank.wire": bank}, Svc: svc,
	})
	if err != nil {
		t.Fatalf("run: %v", err)
	}
	if !report.Passed {
		t.Fatalf("expected pass, got: %v", report)
	}
	if bank.ledgerLen() != 1 {
		t.Fatalf("bank=%d, want 1 (exactly-once held)", bank.ledgerLen())
	}
}

// ── construction-time refusal of bad fault config ────────────────────────

func TestLoseAckRefusesBothConnectorAndTool(t *testing.T) {
	if _, err := LoseAck("bank.wire", "wire", 1.0); err == nil {
		t.Fatalf("expected error when both connector and tool set")
	}
	if _, err := LoseAck("", "", 1.0); err == nil {
		t.Fatalf("expected error when neither connector nor tool set")
	}
}
